### PR TITLE
[WIP] Added graylog as a logging target and integration test

### PR DESCRIPTION
**Pull request by shanewxy**
Added Graylog and it's GELF format as a new logging target.

Related issue: https://github.com/rancher/rancher/issues/14062
Related fluentd PR: https://github.com/rancher/fluentd/pull/28
Related types PR: https://github.com/rancher/types/pull/1067
Related system-charts PR: https://github.com/rancher/system-charts/pull/142

### CODE REVIEW

File: pkg/controllers/user/logging/generator/templatematch.go [hunk 1]
```diff
@@ -217,6 +218,21 @@ var MatchTemplate = `
...
+{{- if eq .CurrentTarget "graylog"}}
+    @type gelf
+	host {{.GraylogTemplateWrap.Host}}
+	port {{.GraylogTemplateWrap.Port}}
+	protocol {{.GraylogConfig.Protocol}}
+	{{- if .GraylogConfig.EnableTLS }}
+	tls true
+	{{end}}
+	{{- if and .GraylogConfig.ClientCert .GraylogConfig.ClientKey}}   
+	tls_options {"cert": "{{.CertFilePrefix}}_client-cert.pem", "key": "{{.CertFilePrefix}}_client-key.pem", "all_ciphers": "true", "no_verify": "true"}
```
Comment (aiwantaozi) [inline pkg/controllers/user/logging/generator/templatematch.go]: configuration `ca` is missed, `no_verity` should use the value from `SSLVerify`

File: pkg/controllers/user/logging/utils/graylog.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	"net"
+
+	"github.com/pkg/errors"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config/dialer"
+)
+
+var (
+	graylogTestData = []byte(`{ "version": "1.1", "host": "rancher", "short_message": "A short message", "level": 5, "_some_info": "foo" }`)
+)
```
Comment (aiwantaozi) [inline pkg/controllers/user/logging/utils/graylog.go]: better to use the same test info as other targets https://github.com/rancher/rancher/blob/master/pkg/controllers/user/logging/utils/testwrap.go#L35

what `_some_info` used for? Can we remove it?

File: tests/integration/suite/test_logging.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    tls_config = {
+        "endpoint": endpoint_host + ":" + endpoint_port,
+        "protocol": "tcp",
+        "enableTls": True,
+        "certificate": "-----BEGIN CERTIFICATE-----\
+                            ----END CERTIFICATE-----",
+        "clientCert": "-----BEGIN CERTIFICATE-----\
+                            ----END CERTIFICATE-----",
+        "clientKey": "-----BEGIN PRIVATE KEY-----\
+                            ----END PRIVATE KEY-----",
```
Comment (aiwantaozi) [inline tests/integration/suite/test_logging.py]: add field SSLVerify

**Comment by shanewxy** [comment]
@aiwantaozi updated

File: pkg/controllers/user/logging/utils/graylog.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+		return errors.Wrapf(err, "couldn't write data to graylog %s", w.Endpoint)
+	}
+
+	// try read to check whether the server close connect already
+	// because can't set read deadline for remote dialer, so if the error is timeout will treat as remote server not close the connection
+	if _, err := readDataWithTimeout(conn); err != nil && err != errReadDataTimeout {
+		return errors.Wrapf(err, "couldn't read data from graylog %s", w.Endpoint)
+	}
+
+	return nil
```
Comment (aiwantaozi) [inline pkg/controllers/user/logging/utils/graylog.go]: This methods is quite similar to syslog, can you write out common methods for them.

Comment (shanewxy) [reply pkg/controllers/user/logging/utils/graylog.go]: Updated

**Comment by aiwantaozi** [comment]
please rebase to resolve conflict

**Comment by shanewxy** [comment]
@aiwantaozi rebased

**Comment by aiwantaozi** [comment]
Can you squash the second and fourth commit to one commit?

**Comment by shanewxy** [comment]
@aiwantaozi rebased

**Pull request closed without merging**